refactor: add subtask register for github plugin

chenggui53 opened the pull request:
### ⚠️ Pre Checklist

> Please complete _ALL_ items in this checklist, and remove before submitting

- [x] I have read through the [Contributing Documentation](https://devlake.apache.org/community/).
- [x] I have added relevant tests.
- [x] I have added relevant documentation.
- [x] I will add labels to the PR, such as `pr-type/bug-fix`, `pr-type/feature-development`, etc.

<!--
Thanks for submitting a pull request!

We appreciate you spending the time to work on these changes.
Please fill out as many sections below as possible.
-->

### Summary
What does this PR do?
like pr https://github.com/apache/incubator-devlake/pull/4896, add subtask register for github

### Does this close any open issues?
Closes #4413

### Screenshots
Include any relevant screenshots here.

### Other Information
Any other information that is important to this PR.

klesh commented on backend/plugins/github/impl/impl.go:
```diff
@@ -95,53 +102,11 @@ func (p Github) Description() string {
-		tasks.ConvertPullRequestCommentsMeta,
-		tasks.ConvertMilestonesMeta,
-		tasks.ConvertAccountsMeta,
+	list, err := subtaskmeta_sorter.NewDependencySorter(tasks.SubTaskMetaList).Sort()
```
Can we reuse the result from https://github.com/apache/incubator-devlake/pull/5411/files#diff-5d9d2b5b08a579490b7eeaba1746905ca45123a978ec531bd584183eec384fbfR51 instead of sorting it twice?

chenggui53 commented:
Hi @klesh, Should i create pr for each plugin to add this register, or do this in one pr ?

klesh commented on backend/plugins/github/tasks/pr_commit_collector.go:
```diff
@@ -44,6 +48,7 @@ var CollectApiPullRequestCommitsMeta = plugin.SubTaskMeta{
 	EnabledByDefault: true,
 	Description:      "Collect PullRequestCommits data from Github api, supports both timeFilter and diffSync.",
 	DomainTypes:      []string{plugin.DOMAIN_TYPE_CROSS, plugin.DOMAIN_TYPE_CODE_REVIEW},
+	Dependencies:     []*plugin.SubTaskMeta{&ExtractApiEventsMeta},
```
I just realize that the `Dependencies` are defined based on the current order which would work but is not accurate.
For example, the `CollectApiPullRequestCommitsMeta` should depend on the `ExtractApiPullRequestMete` instead of `ExtractApiEventsMeta`.
Would you mind updating the code to reflect that? The trick is to identify [which table](https://github.com/apache/incubator-devlake/pull/5411/files#diff-124f6bf67e82ddaafdc0fb6d16ab9ed8afa26c0937a25b259046520ebe1ada4dR79) is required in the collector as the `Input` parameter, and find out the Extractor of the table.

klesh commented:
Hi, @chenggui53 , thanks for asking. I think both are fine for this refactor.
I saw the dependencies are deduced from the Hardcoded Order which is not accurate, would you mind fixing them as well? Maybe we should do it One by One so I can verify the dependency for you.

BTW, Your contribution means a lot to the project, would you like to join us as an Apache Committer?

chenggui53 commented:
> Hi, @chenggui53 , thanks for asking. I think both are fine for this refactor. I saw the dependencies are deduced from the Hardcoded Order which is not accurate, would you mind fixing them as well? Maybe we should do it One by One so I can verify the dependency for you.
> 
> BTW, Your contribution means a lot to the project, would you like to join us as an Apache Committer?

Your comments are very helpful, this register is a step forward, I will fix this part as soon as possible. 
And thank you very much for your invitation, I am very happy to be a committer of this project

klesh commented:
> > Hi, @chenggui53 , thanks for asking. I think both are fine for this refactor. I saw the dependencies are deduced from the Hardcoded Order which is not accurate, would you mind fixing them as well? Maybe we should do it One by One so I can verify the dependency for you.
> > BTW, Your contribution means a lot to the project, would you like to join us as an Apache Committer?
> 
> @klesh Your comments are very helpful, this register is a step forward, I will fix this part as soon as possible. And thank you very much for your invitation, I am very happy to be a committer of this project

Nice, I will nominate you ASAP, it may take a couple of weeks.

klesh commented:
@chenggui53 I thought we agree on "One PR for Plugin" so I can help verify the dependencies are correct. Seems like this PR contains multiple plugins.

chenggui53 commented:
> @chenggui53 I thought we agree on "One PR for Plugin" so I can help verify the dependencies are correct. Seems like this PR contains multiple plugins.  

@klesh sorry, i've modified multiple plugins in this branch for test，it's not ready for review. i will create a new branch for test and request your review when it done

chenggui53 committed:
refactor: add subtask registert for github

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 commented:
hi @klesh , I've done a version, can you help to review it when you have time :)
here are the changes i've done:
1. Add table-based dependency registration and sorting in plugin helper
2. Add register and table dependencies in github plugin
3. To avoid changes in the gitlab plugin, I've kept subtaskMeta dependencies code and i will update it in another new pr.

Some restrictions are also introduced
1. If a subtask task category depends on another task to handle this table, then an temporary table dependency should be added to the dependency of this task to ensure that a single data table is processed before that, which I think is also very understandable to developers.
e.g. like EnrichPullRequestIssuesMeta and ConvertPullRequestIssuesMeta in github plugin, they depends on PullReqeust task to process data in RAW_PULL_REQUEST_TABLE, so i added tool_layer in their table dependency

With that changes, we can process subtask concurrency (like we discussed in pr #4896),  I can submit a new PR after completing the subtask register changes to support concurrent execution of subtasks in plugins.

klesh commented:
> hi @klesh , I've done a version, can you help to review it when you have time :) here are the changes i've done:
> 
> 1. Add table-based dependency registration and sorting in plugin helper
> 2. Add register and table dependencies in github plugin
> 3. To avoid changes in the gitlab plugin, I've kept subtaskMeta dependencies code and i will update it in another new pr.
> 
> Some restrictions are also introduced
> 
> 1. If a subtask task category depends on another task to handle this table, then an temporary table dependency should be added to the dependency of this task to ensure that a single data table is processed before that, which I think is also very understandable to developers.
>    e.g. like EnrichPullRequestIssuesMeta and ConvertPullRequestIssuesMeta in github plugin, they depends on PullReqeust task to process data in RAW_PULL_REQUEST_TABLE, so i added tool_layer in their table dependency
> 
> With that changes, we can process subtask concurrency (like we discussed in pr #4896), I can submit a new PR after completing the subtask register changes to support concurrent execution of subtasks in plugins.

OMG, you have finished the table dependency sorting? way to go, man.
Will look ASAP, however, I'm kind of hands full at the moment.
@keon94 @CamilleTeruel Please take a look at it as well.

klesh reviewed (commented):
Not sure if I understand it correctly, seems like subtasks are declaring only the dependency , how does it work? I believe we need both **dependency** and **product** tables to connect those **subtasks** correctly.

chenggui53 commented:
> Not sure if I understand it correctly, seems like subtasks are declaring only the dependency , how does it work? I believe we need both **dependency** and **product** tables to connect those **subtasks** correctly.

It seems that I have a misunderstanding here. Let me have a look again 😅

klesh commented:
I tried to draw a  DAG to depict how the data flow might look, I hope that gives you the idea or the problem, feel free to ask me anything

![2023-06-28 2023-06-28 11 34 38 excalidraw](https://github.com/apache/incubator-devlake/assets/61080/f6fde36d-3d70-4434-8232-fca8052986c8)

chenggui53 commented:
> I tried to draw a DAG to depict how the data flow might look, I hope that gives you the idea or the problem, feel free to ask me anything
> 
> ![2023-06-28 2023-06-28 11 34 38 excalidraw](https://user-images.githubusercontent.com/61080/249343697-f6fde36d-3d70-4434-8232-fca8052986c8.svg)

This graph is very helpful, I will improve it ASAP 😀

keon94 commented on backend/helpers/pluginhelper/subtaskmeta_sorter/subtask.go:
```diff
@@ -32,11 +32,11 @@ func NewDependencySorter(metas []*plugin.SubTaskMeta) SubTaskMetaSorter {
 }
 
 // stable topological sort
-func topologicalSort(metas []*plugin.SubTaskMeta) ([]plugin.SubTaskMeta, error) {
```
Our convention is to use errors.Error (our own impl) in our code instead of the base Go error interface

keon94 commented on backend/helpers/pluginhelper/subtaskmeta_sorter/subtask.go:
```diff
@@ -60,59 +60,21 @@ func topologicalSort(metas []*plugin.SubTaskMeta) ([]plugin.SubTaskMeta, error)
+	for _, item := range orderStrList {
+		value, ok := nameMetaMap[item]
+		if !ok {
+			return nil, fmt.Errorf("illeagal subtaskmeta detected %s", item)
```
typo: "illegal"

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 commented:
> > > Hi, @chenggui53 , thanks for asking. I think both are fine for this refactor. I saw the dependencies are deduced from the Hardcoded Order which is not accurate, would you mind fixing them as well? Maybe we should do it One by One so I can verify the dependency for you.
> > > BTW, Your contribution means a lot to the project, would you like to join us as an Apache Committer?
> > 
> > 
> > @klesh Your comments are very helpful, this register is a step forward, I will fix this part as soon as possible. And thank you very much for your invitation, I am very happy to be a committer of this project
> 
> Nice, I will nominate you ASAP, it may take a couple of weeks.

Hi @klesh , I haven't received any relevant emails or messages until now. In order to avoid any notifications that I missed, I would like to ask if this is normal. 😁

klesh commented:
> > > > Hi, @chenggui53 , thanks for asking. I think both are fine for this refactor. I saw the dependencies are deduced from the Hardcoded Order which is not accurate, would you mind fixing them as well? Maybe we should do it One by One so I can verify the dependency for you.
> > > > BTW, Your contribution means a lot to the project, would you like to join us as an Apache Committer?
> > > 
> > > 
> > > @klesh Your comments are very helpful, this register is a step forward, I will fix this part as soon as possible. And thank you very much for your invitation, I am very happy to be a committer of this project
> > 
> > 
> > Nice, I will nominate you ASAP, it may take a couple of weeks.
> 
> Hi @klesh , I haven't received any relevant emails or messages until now. In order to avoid any notifications that I missed, I would like to ask if this is normal. 😁

 Yes, there is one ongoing vote in the queue, we have to wait for that to avoid confusing our mentors. Will be voting for you when it closed.

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask meta register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: Update remote.go

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: update subtask register

chenggui53 commented:
Hi @klesh , something is not clear for me, can you help to check why github.repo_converter set RAW_REPOSITORIES_TABLE  in RawDataSubTaskArgs.Table, and set GithubRepo input for Convert function. I only see RAW_REPOSITORIES_TABLE used once in here,  and can't figure out the releation between RAW_REPOSITORIES_TABLE and GithubRepo.

klesh commented:
> Hi @klesh , something is not clear for me, can you help to check why github.repo_converter set RAW_REPOSITORIES_TABLE in RawDataSubTaskArgs.Table, and set GithubRepo input for Convert function. I only see RAW_REPOSITORIES_TABLE used once in here, and can't figure out the releation between RAW_REPOSITORIES_TABLE and GithubRepo.

Aah, good question, we used to collect repo information from the API and store it to the raw data table.
However, this behavior has been deprecated recently and users would create the tool layer record by calling this API https://github.com/apache/incubator-devlake/blob/990c19ded0f6e23f60aa05cd15bb6bc3e9e66c95/backend/plugins/github/impl/impl.go#L234

In general, no more raw tables for scopes (GitHub repo, jira board, etc.)

chenggui53 committed:
refactor: update subtask register

klesh commented on backend/helpers/pluginhelper/subtaskmeta/sorter/table.go:
```diff
@@ -0,0 +1,104 @@
+		// check if subtask product same tables, which may cause cyclic error
+		for _, productItem := range item.ProductTables {
+			if contains(productList, productItem) {
+				return nil, errors.Convert(fmt.Errorf("duplicate product detected %s", productItem))
```
It doesn't seem right because **records in a table may come from multiple subtasks**, for example
the jira issue extractor [produces `Changelog`](https://github.com/apache/incubator-devlake/blob/927db42404f393557b26982761c70b06f462d0da/backend/plugins/jira/tasks/issue_extractor.go#L154) while [changelog collector](https://github.com/apache/incubator-devlake/blob/15afa46cf82ce479fb46757734636cf787faff69/backend/plugins/jira/tasks/issue_changelog_collector.go) + [changelog extractor](https://github.com/apache/incubator-devlake/blob/2d08ffaedc9debcb0be367bb84fa18592bb06cec/backend/plugins/jira/tasks/issue_changelog_extractor.go) also producing records to the same table due to Performance Consideration: If we could collect all changelogs from the issues, then we don't need to collect them from the Changelog API which is very very slow.

Maybe we could just deduce and update the `Dependencies` for `SubtaskMeta` based on `ProductTables` and `DependencyTables` and reuse the existing `SubTaskMetaSorter`. 
1. TableSort does nothing more than figure out what `Dependencies` should be
2. Sorting and Cyclic Detection can be done in `SubTaskMetaSorter` and one place only

chenggui53 replied:
In fact, when I wrote here, I wondered whether there would be a sequence and conflict between different tasks operating on the same table, but unfortunately, I didn't think about this issue carefully and ask you for confirmation in time 😅

klesh replied:
No problem, it happens.

chenggui53 commented on backend/helpers/pluginhelper/subtaskmeta/sorter/table.go:
```diff
@@ -0,0 +1,74 @@
+	return dependenciesTopologicalSort(constructedMetas)
+}
+
+func constructDependenciesByTable(metas []*plugin.SubTaskMeta) []*plugin.SubTaskMeta {
```
Hi @klesh , I'm not sure if this is what you mentioned before. Can you help review it.

chenggui53 commented on backend/plugins/github/tasks/commit_convertor.go:
```diff
@@ -64,7 +77,7 @@ func ConvertCommits(taskCtx plugin.SubTaskContext) errors.Error {
 				Name:         data.Options.Name,
 			},
-			Table: RAW_COMMENTS_TABLE,
+			Table: RAW_COMMIT_TABLE,
```
Hi @klesh , i'm not sure if it's a typo here, commit converter uses a raw_comments_talbe instead of a raw_commit_table. Can you help to check this.

klesh replied:
You are correct, definitely a typo for sure.

chenggui53 commented on backend/helpers/pluginhelper/subtaskmeta/sorter/table.go:
```diff
@@ -0,0 +1,74 @@
+	return dependenciesTopologicalSort(constructedMetas)
+}
+
+func constructDependenciesByTable(metas []*plugin.SubTaskMeta) []*plugin.SubTaskMeta {
```
Hi @klesh , does this function match what you mentioned earlier?

klesh replied:
I believe so. Good job.

chenggui53 committed:
refactor: add subtask register

klesh commented on backend/helpers/pluginhelper/subtaskmeta/sorter/table.go:
```diff
@@ -0,0 +1,74 @@
+	}
+	// construct meta dependencies by meta.TableDependencies
+	for _, metaItem := range metas {
+		for _, tableItem := range metaItem.DependencyTables {
```
It would be better to create a temporary `map` for each `metaItem` and reuse it for deduplication instead of creating one on each `append`

chenggui53 replied:
thanks, i will update it later

chenggui53 replied:
Thanks, I will update later.

klesh commented on backend/helpers/pluginhelper/subtaskmeta/sorter/table_test.go:
```diff
@@ -0,0 +1,81 @@
+	pluginA := plugin.SubTaskMeta{
+		Name:             "A",
+		DependencyTables: []string{},
+		ProductTables:    []string{"_TOOL_TEST_TABLE"},
```
It would be better to simulate the situation I described in the previous review that multiple subtasks might produce records for the same table

chenggui53 committed:
refactor: add subtask register

chenggui53 committed:
refactor: add subtask register

klesh commented:
@chenggui53 Seems ready to merge, am I right?

chenggui53 commented:
> @chenggui53 Seems ready to merge, am I right?

@klesh Yes, I think so.

klesh reviewed (approved):
LGTM
